core/src/main/java/org/apache/druid/timeline/partition/PartitionHolder.java

```diff
@@ -22,72 +22,62 @@
 import com.google.common.collect.Iterables;
 import com.google.common.collect.Iterators;
 
+import javax.annotation.Nullable;
 import java.util.Iterator;
 import java.util.List;
-import java.util.SortedSet;
 import java.util.Spliterator;
-import java.util.TreeSet;
+import java.util.TreeMap;
 
 /**
  * An object that clumps together multiple other objects which each represent a shard of some space.
  */
 public class PartitionHolder<T> implements Iterable<PartitionChunk<T>>
 {
-  private final TreeSet<PartitionChunk<T>> holderSet;
+  private final TreeMap<PartitionChunk<T>, PartitionChunk<T>> holderMap;
 
   public PartitionHolder(PartitionChunk<T> initialChunk)
   {
-    this.holderSet = new TreeSet<>();
+    this.holderMap = new TreeMap<>();
     add(initialChunk);
   }
 
   public PartitionHolder(List<PartitionChunk<T>> initialChunks)
   {
-    this.holderSet = new TreeSet<>();
+    this.holderMap = new TreeMap<>();
     for (PartitionChunk<T> chunk : initialChunks) {
       add(chunk);
     }
   }
 
-  public PartitionHolder(PartitionHolder partitionHolder)
+  public PartitionHolder(PartitionHolder<T> partitionHolder)
   {
-    this.holderSet = new TreeSet<>();
-    this.holderSet.addAll(partitionHolder.holderSet);
+    this.holderMap = new TreeMap<>();
+    this.holderMap.putAll(partitionHolder.holderMap);
   }
 
-  public void add(PartitionChunk<T> chunk)
+  public boolean add(PartitionChunk<T> chunk)
   {
-    holderSet.add(chunk);
+    return holderMap.putIfAbsent(chunk, chunk) == null;
```

Comment:
I suppose `TreeMap` was chosen because of the method `putIfAbsent`?
Anyway, I suspect we don't call `add(chunk)` with the same chunk twice, or do I miss something?

Reply:
I just replaced `TreeSet` with `TreeMap`, not sure how does it relate to `putIfAbsent()`.

> Anyway, I suspect we don't call add(chunk) with the same chunk twice, or do I miss something?

I don't know. This is unprovable from the point of view of the `VersionedIntervalTimeline` abstraction, which may contain any objects and anybody may call its `add()` method with anything. So `VersionedIntervalTimeline` assumes that objects can be repeated and increments `numObjects` counter only when `PartitionHolder.add()` returns true.